source/common/network/buffered_io_socket_handle_impl.h

```diff
@@ -0,0 +1,163 @@
+  Api::IoCallUint64Result read(Buffer::Instance& buffer, uint64_t max_length) override;
+
+  Api::IoCallUint64Result writev(const Buffer::RawSlice* slices, uint64_t num_slice) override;
+
+  Api::IoCallUint64Result write(Buffer::Instance& buffer) override;
+
+  Api::IoCallUint64Result sendmsg(const Buffer::RawSlice* slices, uint64_t num_slice, int flags,
+                                  const Address::Ip* self_ip,
+                                  const Address::Instance& peer_address) override;
+
+  Api::IoCallUint64Result recvmsg(Buffer::RawSlice* slices, const uint64_t num_slice,
+                                  uint32_t self_port, RecvMsgOutput& output) override;
+
+  Api::IoCallUint64Result recvmmsg(RawSliceArrays& slices, uint32_t self_port,
+                                   RecvMsgOutput& output) override;
+  Api::IoCallUint64Result recv(void* buffer, size_t length, int flags) override;
+
+  bool supportsMmsg() const override;
+  bool supportsUdpGro() const override;
+
+  Api::SysCallIntResult bind(Address::InstanceConstSharedPtr address) override;
+  Api::SysCallIntResult listen(int backlog) override;
+  IoHandlePtr accept(struct sockaddr* addr, socklen_t* addrlen) override;
+  Api::SysCallIntResult connect(Address::InstanceConstSharedPtr address) override;
+  Api::SysCallIntResult setOption(int level, int optname, const void* optval,
+                                  socklen_t optlen) override;
+  Api::SysCallIntResult getOption(int level, int optname, void* optval, socklen_t* optlen) override;
+  Api::SysCallIntResult setBlocking(bool blocking) override;
+  absl::optional<int> domain() override;
+  Address::InstanceConstSharedPtr localAddress() override;
+  Address::InstanceConstSharedPtr peerAddress() override;
+  Event::FileEventPtr createFileEvent(Event::Dispatcher& dispatcher, Event::FileReadyCb cb,
+                                      Event::FileTriggerType trigger, uint32_t events) override;
+  Api::SysCallIntResult shutdown(int how) override;
+  absl::optional<std::chrono::milliseconds> lastRoundTripTime() override { return absl::nullopt; }
+
+  Buffer::WatermarkBuffer& getBufferForTest() { return owned_buffer_; }
+
+  void scheduleNextEvent() {
+    // It's possible there is no pending file event so as no io_callback.
+    if (io_callback_) {
+      ENVOY_LOG(trace, "Schedule IO callback on {}", static_cast<void*>(this));
+      io_callback_->scheduleCallbackNextIteration();
+    }
+  }
+
+  void setWritablePeer(WritablePeer* writable_peer) {
+    // Swapping writable peer is undefined behavior.
+    ASSERT(!writable_peer_);
+    ASSERT(!write_shutdown_);
+    writable_peer_ = writable_peer;
+  }
+
+  // WritablePeer
+  void setWriteEnd() override { read_end_stream_ = true; }
+  bool isWriteEndSet() override { return read_end_stream_; }
+  void maybeSetNewData() override {
+    ENVOY_LOG(trace, "{} on socket {}", __FUNCTION__, static_cast<void*>(this));
+    scheduleNextEvent();
+  }
+  void onPeerDestroy() override {
```

Comment:
Any chance we can have writable peer on construction?  If so it can be a reference. 
If it stays a pointer, having a bool and nullptr seem redundant.

Reply:
> Any chance we can have writable peer on construction? If so it can be a reference.

It cannot be reference because we are constructing two `BufferedIoSocketHandleImpl` and references each other with the pointer. It's error prone so I introduced the factory to avoid manually construct the 

> If it stays a pointer, having a bool and nullptr seem redundant.

The flag you are referring to is `write_shutdown_`, right? The flag is not quite the same as `peer == nullptr`. `write_shutdown_` is to track the call of `BufferedIoSocketHandleImpl::shutdown(int how)`.  So after `shutdown` but before `close`, this `peer` must be remain valid.